### PR TITLE
server: close session if old capture dead.

### LINKED ISSUE #2388: Capture list increased 1 unexpected after all PD restarting
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

- 2x capture:
  Starting component `cdc`: /root/.tiup/components/cdc/v5.1.0/cdc cli capture list --pd=http://172.16.6.24:2379
[
  {
    "id": "3378b726-26cb-4963-8280-3ee679024a76",
    "is-owner": false,
    "address": "172.16.6.32:8300"
  },
  {
    "id": "d2600da4-cdc4-4420-84a4-e57e826ffbc7",
    "is-owner": true,
    "address": "172.16.6.31:8300"
  }
]

- Restart all PD:  $ tiup cluster restart 360UP -R pd
- Check capture list

2. What did you expect to see?

- 2x capture, and their status is normal.

3. What did you see instead?

- 3x capture list unexpected in a short time after PD restarting:
Starting component `cdc`: /root/.tiup/components/cdc/v5.1.0/cdc cli capture list --pd=http://172.16.6.24:2379
[
  {
    "id": "2b0211bd-f3fc-4551-b4d1-a5c6bba5818e",
    "is-owner": false,
    "address": "172.16.6.32:8300"
  },
  {
    "id": "3378b726-26cb-4963-8280-3ee679024a76",
    "is-owner": false,
    "address": "172.16.6.32:8300"
  },
  {
    "id": "d2600da4-cdc4-4420-84a4-e57e826ffbc7",
    "is-owner": true,
    "address": "172.16.6.31:8300"
  }
]
Waiting for several seconds,  2x capture list as expected:
Starting component `cdc`: /root/.tiup/components/cdc/v5.1.0/cdc cli capture list --pd=http://172.16.6.24:2379
[
  {
    "id": "2b0211bd-f3fc-4551-b4d1-a5c6bba5818e",
    "is-owner": true,
    "address": "172.16.6.32:8300"
  },
  {
    "id": "4b8e6f29-847d-4caf-bc7a-ea8cba317a28",
    "is-owner": false,
    "address": "172.16.6.31:8300"
  }
]

4. Versions of the cluster

    - Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

        ```
        4.0.14
        ```

    - TiCDC version (execute `cdc version`):

        ```
        [release-version=v4.0.14] [git-hash=5a7851967f686da896b45acd3f3e968bfe53d6bd] [git-branch=heads/refs/tags/v4.0.14]
        ```

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve? 

This PR close #2388 

### What is changed and how it works?

In the old capture implementation, the Etcd session is not closed, which may cause multiple capture info registered in etcd, because outdated capture's info is not deleted in time.

This PR only works on branches that use old capture implementation.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

the same as #2388. no more ghost captures after this fix.

Code changes

Side effects

Related changes

### Release note 

```release-note
fix outdated capture info may appear in capture list command
```

### CODE REVIEW

**Comment by liuzix** [comment]
Is the session leaked? Or will the session be dropped eventually?

File: cdc/capture.go [hunk 1]
```diff
@@ -143,6 +143,13 @@ func NewCapture(
 func (c *Capture) Run(ctx context.Context) (err error) {
 	ctx, cancel := context.WithCancel(ctx)
 	// TODO: we'd better to add some wait mechanism to ensure no routine is blocked
+	defer func() {
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
Comment (amyangfei) [inline cdc/capture.go:147]: Should the timeout larger than `CaptureSessionTTL`

File: cdc/server.go [hunk 2]
```diff
@@ -298,12 +298,22 @@ func (s *Server) etcdHealthChecker(ctx context.Context) error {
 func (s *Server) run(ctx context.Context) (err error) {
 	if !config.NewReplicaImpl {
 		kvStorage := util.KVStorageFromCtx(ctx)
+		// if s.capture != nil && s.capture.session != nil {
+		// 	s.capture.session.Close() //nolint:errcheck
+		// }
```
Comment (amyangfei) [inline cdc/server.go:303]: Why adding this logic, but commented out

Comment (3AceShowHand) [reply cdc/server.go:303]: sry, it's still testing.

**Comment by 3AceShowHand** [comment]
> Is the session leaked? Or will the session be dropped eventually?

Yes, the old capture's session is not properly closed, But I think It will be dropped after the keepalive timeout because ghost capture will disappear a few seconds later.

**Comment by overvenus** [comment]
Should it be cherry-picked to release-5.0?

**Comment by 3AceShowHand** [comment]
> Should it be cherry-picked to release-5.0?

yes, it also uses old capture implementation.

File: cdc/server.go [hunk 3]
```diff
@@ -298,12 +298,25 @@ func (s *Server) etcdHealthChecker(ctx context.Context) error {
...
+		}
 		capture, err := NewCapture(ctx, s.pdEndpoints, s.pdClient, kvStorage)
 		if err != nil {
 			return err
 		}
 		s.capture = capture
 		s.etcdClient = &capture.etcdClient
+		conf := config.GetGlobalServerConfig()
+		defer func() {
+			timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Duration(conf.CaptureSessionTTL+5)*time.Second)
```
Comment (amyangfei) [inline cdc/server.go:314]:
```suggestion
			timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Duration(conf.CaptureSessionTTL)*time.Second)
```
Is `time.Duration(conf.CaptureSessionTTL)` enough?

**Review by liuzix — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Pull request merged**